hudi-common/src/main/java/org/apache/hudi/common/schema/HoodieSchema.java

```diff
@@ -0,0 +1,944 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.schema;
+
+import org.apache.hudi.avro.HoodieAvroUtils;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.ValidationUtils;
+import org.apache.hudi.exception.HoodieAvroSchemaException;
+
+import org.apache.avro.JsonProperties;
+import org.apache.avro.Schema;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Wrapper class for Avro Schema that provides Hudi-specific schema functionality
+ * while maintaining binary compatibility with Avro.
+ *
+ * <p>This class serves as the primary interface for schema operations within Hudi.
+ * It encapsulates an Avro Schema and provides a consistent, type-safe API while
+ * maintaining full compatibility with existing Avro-based code.</p>
+ *
+ * <p>Key features:
+ * <ul>
+ *   <li>Binary compatibility with Avro Schema</li>
+ *   <li>Type-safe field access through HoodieSchemaField</li>
+ *   <li>Support for all Avro schema types and operations</li>
+ *   <li>Consistent error handling using Hudi exceptions</li>
+ *   <li>Integration with Hudi's Option type for null safety</li>
+ * </ul></p>
+ *
+ * <p>Usage examples:
+ * <pre>{@code
+ * // Create from JSON
+ * HoodieSchema schema = HoodieSchema.parse(jsonSchemaString);
+ *
+ * // Create primitive schemas
+ * HoodieSchema stringSchema = HoodieSchema.create(HoodieSchemaType.STRING);
+ *
+ * // Access schema properties
+ * HoodieSchemaType type = schema.getType();
+ * List<HoodieSchemaField> fields = schema.getFields();
+ * Option<HoodieSchemaField> field = schema.getField("fieldName");
+ *
+ * // Convert back to Avro for compatibility
+ * Schema avroSchema = schema.getAvroSchema();
+ * }</pre></p>
+ *
+ * @since 1.2.0
+ */
+public class HoodieSchema implements Serializable {
```

Comment:
Once this class `HoodieSchema`, `HoodieSchemaField` and other schema-related classes are integrated, we should mark these `@PublicAPIClass` as I imagine the new schema system will also be used by other engine integration outside Hudi repo such as Presto and Trino.